Add all course events to user calendar

omerhabib26 opened the pull request:
### Description

[LEARNER-8320](https://openedx.atlassian.net/browse/LEARNER-8320), [LEARNER-8381](https://openedx.atlassian.net/browse/LEARNER-8381)

- Integrate Mobile App calendar with edX App
- Ask Permission to access the calendar app
- show alert dialogs based on success/failure
- Added course events to the local mobile app calendar
- Add a separate calendar for every course
- Added [Analytics](https://docs.google.com/spreadsheets/d/1SCwq_4vFT9OjxonMpx9WiNvhkcDSGWWlJK_-I4w2_BA/edit#gid=1442327496)
- Ref Doc: https://developer.android.com/guide/topics/providers/calendar-provider

<img src="https://user-images.githubusercontent.com/30689349/121079391-7d69d400-c7f3-11eb-837c-0e865bb2e0a6.png" height="360"/> <img src="https://user-images.githubusercontent.com/30689349/121079382-7c38a700-c7f3-11eb-9987-23cd1f5b2e98.png" height="360"/> <img src="https://user-images.githubusercontent.com/30689349/121079889-2d3f4180-c7f4-11eb-8c87-7d24c58223a3.png" height="360"/>  <img src="https://user-images.githubusercontent.com/30689349/121079387-7cd13d80-c7f3-11eb-83ac-3fac76241a42.png" height="360"/> <img src="https://user-images.githubusercontent.com/30689349/121079364-7773f300-c7f3-11eb-8762-e24407e7547a.png" height="360"/>  <img src="https://user-images.githubusercontent.com/30689349/121079874-29132400-c7f4-11eb-9bb8-f56f5439155f.png" height="360"/>

PS: `Currently we are adding course calendar as a separate entity not with the edX provide email address`

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_course_dates_page.xml:
```diff
@@ -33,6 +33,87 @@
+                                android:layout_height="wrap_content"
+                                android:orientation="horizontal">
+
+                                <TextView
```
should set android:gravity="center_vertical".

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_course_dates_page.xml:
```diff
@@ -33,6 +33,87 @@
+                            android:textSize="@dimen/edx_small" />
+                    </LinearLayout>
+
+                    <androidx.cardview.widget.CardView
```
set the top margin for the card to `24dp`

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -167,6 +193,72 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+        val title: String = ResourceUtil.getFormattedString(resources, R.string.title_add_course_calendar, AppConstants.COURSE_NAME, courseName).toString()
+        val message: String = ResourceUtil.getFormattedString(resources, R.string.message_after_course_calendar_added, AppConstants.COURSE_NAME, courseName).toString()
+
+        AlertDialogFragment.newInstance(title, message, getString(R.string.label_done),
```
There is no title as per deisngs.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -167,6 +193,72 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+        AlertDialogFragment.newInstance(title, message, getString(R.string.label_ok),
+                { dialogInterface: DialogInterface, i: Int ->
+                    environment.analyticsRegistry.trackCourseDatesCalendarAddDates(courseData.courseId, courseData.mode, isSelfPaced, true)
+                    insertCalendarEvent()
```
It should be async, cuz this moves the app in a wait busy state for lower-end devices like Nexus-5.

farhan-arshad-dev reviewed (changes requested):
- Please add the screenshots for the UI changes.
- Application is crashing on Android 9, Vivo Mobile.
- Crashing on Android 5.0 when user toggled on calendar sync switch and Don’t allow on the permission dialog.
- Crashing on Android 5.0 when user toggled on calendar sync switch and Click ok on the permission dialog, and Deny the calendar permission.
- Dialogs button should be with bold text.
- What would be the finalized Calendar Switch `TOGGLE OFF STYLE` & `TOGGLE ON STYLE`.
- The dialog message text color should be `707070` to match the design.
- The old dates banner should be updated as per new proposed designs: https://www.figma.com/file/4N4hsN0gXjG2GFw0WxMKCd/Calendar-sync?node-id=72%3A512

omerhabib26 reviewed (commented):
- Crashes Fixed
> * Application is crashing on Android 9, Vivo Mobile.
> * Crashing on Android 5.0 when user toggled on calendar sync switch and Don’t allow on the permission dialog.
> * Crashing on Android 5.0 when user toggled on calendar sync switch and Click ok on the permission dialog, and Deny the calendar permission.
- Button Text Updated
> * Dialogs button should be with bold text.
- As Per Discussion, native toggle button will be used
> * What would be the finalized Calendar Switch `TOGGLE OFF STYLE` & `TOGGLE ON STYLE`.
- As per sigma designs dialog text color is black with different transparency levels, so I didn't update them cox it will effect all the dialogs in the app.
> * The dialog message text color should be `707070` to match the design.
- Updated
> * The old dates banner should be updated as per new proposed designs: https://www.figma.com/file/4N4hsN0gXjG2GFw0WxMKCd/Calendar-sync?node-id=72%3A512

farhan-arshad-dev commented:
@omerhabib26 

- Calendar Sync switch is not visible on Android 9.
- Do we need to update the designs of the course dates banner for all screens.
- Need to update the calendar sync toggle as per designs: https://www.figma.com/file/4N4hsN0gXjG2GFw0WxMKCd/Calendar-sync?node-id=72%3A512
- Increase the line space of the course dates/calendar sync banner, and need to match the font size.
- I’m not sure but on an emulator, the event is not removed from the calendar is not toggle off the calendar sync.
- Increase the line spacing of the dialog too (if possible while using native functionally).

omerhabib26 commented:
> @omerhabib26
Please verify the Firebase RemoteConfig Response
> * Calendar Sync switch is not visible on Android 9.
Updated the designs as per discussion with @mumer92
> * Do we need to update the designs of the course dates banner for all screens.
Updated
> * Need to update the calendar sync toggle as per designs: https://www.figma.com/file/4N4hsN0gXjG2GFw0WxMKCd/Calendar-sync?node-id=72%3A512
Updated
> * Increase the line space of the course dates/calendar sync banner, and need to match the font size.
Please verify it as it's working fine on my side
> * I’m not sure but on an emulator, the event is not removed from the calendar is not toggle off the calendar sync.
Updated
> * Increase the line spacing of the dialog too (if possible while using native functionally).

farhan-arshad-dev commented on OpenEdXMobile/res/values/styles.xml:
```diff
@@ -579,4 +580,10 @@
     </style>
+
+    <style name="sync.Calendar.Switch" parent="Theme.AppCompat.Light">
+        <item name="colorControlActivated">@color/neutralBlackT</item>
```
as per design, it should be `primaryDarkColor(002121)`

omerhabib26 replied:
Done

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/model/course/CourseDatesBannerInfo.kt:
```diff
@@ -16,6 +16,10 @@ data class CourseDatesBannerInfo(
 
+    // TODO currently setting it to true for development purpose
+    fun showCalendarSync(): Boolean = true
+//            upgradeToGraded() || showBannerInfo()
```
Reminder.

omerhabib26 replied:
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -167,6 +194,87 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+
+    private fun calendarAddedSuccessDialog() {
+        isCalendarExist = true
+        val title: String = ResourceUtil.getFormattedString(resources, R.string.title_add_course_calendar, AppConstants.COURSE_NAME, courseName).toString()
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CourseDateUtil.kt:
```diff
@@ -16,44 +18,50 @@ import org.edx.mobile.module.analytics.AnalyticsRegistry
 object CourseDateUtil {
 
-    fun setupCourseDatesBanner(view: View, courseId: String, enrollmentMode: String, isSelfPaced: Boolean,
+    fun setupCourseDatesBanner(view: View, isCourseDatePage: Boolean, courseId: String, enrollmentMode: String, isSelfPaced: Boolean,
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/ConfigUtil.kt:
```diff
@@ -126,6 +129,32 @@ class ConfigUtil {
+                val firebaseRemoteConfig = FirebaseRemoteConfig.getInstance()
+                firebaseRemoteConfig.fetchAndActivate().addOnCompleteListener {
+                    try {
+                        val response: JSONObject = JSONObject(firebaseRemoteConfig
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -1,37 +1,50 @@
 import android.view.ViewGroup
+import android.widget.Toast
 import androidx.databinding.DataBindingUtil
+import androidx.databinding.library.baseAdapters.BuildConfig.VERSION_NAME
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/model/course/CourseDatesBannerInfo.kt:
```diff
@@ -9,8 +9,8 @@ data class CourseDatesBannerInfo(
-        upgradeToGraded() -> CourseBannerType.UPGRADE_TO_GRADED
-        upgradeToReset() -> CourseBannerType.UPGRADE_TO_RESET
+        upgradeToGraded() -> CourseBannerType.INFO_BANNER
+        upgradeToReset() -> CourseBannerType.INFO_BANNER
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -27,6 +27,10 @@
 
+    public static final String PLATFORM_NAME = "platform_name";
+    public static final String COURSE_NAME = "course_name";
+    public static final String KEY_ANDROID = "android";
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/dialog/AlertDialogFragment.java:
```diff
@@ -176,9 +180,30 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
         return alertDialog;
     }
 
+    @Override
+    public void onViewCreated(View view, Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+    }
```
removed

omerhabib26 commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -238,27 +238,56 @@
-    <!-- Message Course Dates Banner Upgrade to Reset -->
-    <string name="course_dates_banner_upgrade_to_reset"><b>You are auditing this course,</b>
+    <!-- Message Course Dashboard Banner Upgrade to Reset -->
+    <string name="course_dashboard_banner_upgrade_to_reset"><b>You are auditing this course,</b>>
```
updated

omerhabib26 commented on OpenEdXMobile/res/layout/fragment_course_dates_page.xml:
```diff
@@ -33,6 +33,65 @@
+                                    app:trackTint="@color/switch_track_color" />
+                            </LinearLayout>
+
+                            <TextView
```
added

omerhabib26 commented on OpenEdXMobile/res/drawable/ic_sync_alt.xml:
```diff
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="26dp"
+    android:height="26dp"
```
updated

omerhabib26 commented on OpenEdXMobile/res/layout/fragment_course_dates_page.xml:
```diff
@@ -33,6 +33,65 @@
+                                <androidx.appcompat.widget.SwitchCompat
+                                    android:id="@+id/switch_sync"
+                                    android:layout_width="wrap_content"
+                                    android:layout_height="@dimen/edx_double_margin"
```
done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -196,4 +316,115 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+            val values = ContentValues().apply {
+                put(CalendarContract.Events.DTSTART, startMillis)
+                put(CalendarContract.Events.DTEND, endMillis)
+                put(CalendarContract.Events.TITLE, "Assignment Due: ${courseData.course.name}")
```
Done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -196,4 +316,115 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+
+    @Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+    private fun createCalendar(accountName: String): Long {
+        val v = ContentValues()
```
done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -196,4 +316,115 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+            return;
+        }
+
+        val courseDates = viewModel.courseDates.value
+        courseDates?.courseDateBlocks?.forEach { courseDate ->
+            val date = courseDate.getDateCalendar()
+            val startMillis: Long = Calendar.getInstance().run {
+                set(date.get(Calendar.YEAR), date.get(Calendar.MONTH), date.get(Calendar.DAY_OF_MONTH), date.get(Calendar.HOUR_OF_DAY) - 1, date.get(Calendar.MINUTE))
+                timeInMillis
+            }
+            val endMillis: Long = Calendar.getInstance().run {
+                set(date.get(Calendar.YEAR), date.get(Calendar.MONTH), date.get(Calendar.DAY_OF_MONTH), date.get(Calendar.HOUR_OF_DAY), date.get(Calendar.MINUTE))
+                timeInMillis
+            }
```
Added

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     * @param canAddDates True if the user confirmed to allow the events to be added on calendar otherwise False
+     */
+    void trackCourseDatesCalendarAddDates(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced, boolean canAddDates);
+
+    /**
+     * Track if the events added on calendar successfully
+     *
+     * @param courseId    Id of the course
+     * @param userType    Whether the Account type of the user is audit or verified
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     */
+    void trackCourseDatesCalendarAddedSuccessfully(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced);
+
+    /**
+     * Track if the events removed from calendar successfully
+     *
+     * @param courseId    Id of the course
+     * @param userType    Whether the Account type of the user is audit or verified
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     */
+    void trackCourseDatesCalendarRemovedSuccessfully(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced);
+
+    /**
+     * Track the user tapped on ViewEvents on Success dialog of the events added on calendar
+     *
+     * @param courseId    Id of the course
+     * @param userType    Whether the Account type of the user is audit or verified
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     */
+    void trackCourseDatesViewCalendar(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced);
+
+    /**
+     * Track the user tapped on Done on Success dialog of the events added on calendar
+     *
+     * @param courseId    Id of the course
+     * @param userType    Whether the Account type of the user is audit or verified
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     */
+    void trackCourseDatesCalendarConfirmSuccess(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced);
```
IMO, we have to create only one method to track calendar events name `trackCalendarEvent`. cuz all the param are the same in calendar events.

omerhabib26 replied:
updated

mumer92 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -175,12 +278,24 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
     private fun initDatesBanner(courseBannerInfo: CourseBannerInfoModel?) {
         if (courseBannerInfo == null || courseBannerInfo.hasEnded) {
             binding.banner.containerLayout.visibility = View.GONE
+            binding.syncCalendarContainer.visibility = View.GONE
```
Currently Calendar Sync option will be available for self paced courses, so if a course is not self paced, please hide Calendar sync option and let header as it is.

omerhabib26 replied:
noted

mumer92 commented on OpenEdXMobile/src/main/java/org/edx/mobile/model/course/CourseDatesBannerInfo.kt:
```diff
@@ -16,6 +16,8 @@ data class CourseDatesBannerInfo(
         else -> CourseBannerType.BLANK
     }
 
+    fun showCalendarSync(): Boolean = upgradeToGraded() || showBannerInfo()
```
Please check for self paced course before showing Calendar Sync option.

omerhabib26 replied:
we just need to check for self paced for now?

mumer92 replied:
Yes.

omerhabib26 commented on OpenEdXMobile/res/layout/layout_course_dates_banner.xml:
```diff
@@ -6,17 +6,31 @@
         android:layout_height="wrap_content"
-        android:background="@color/infoXXLight"
         android:orientation="vertical"
+        android:background="@color/infoXXLight"
```
done

omerhabib26 commented on OpenEdXMobile/res/values/dimens.xml:
```diff
@@ -266,6 +266,8 @@
     <dimen name="shadow_view_height">4dp</dimen>
     <dimen name="container_padding">12dp</dimen>
     <dimen name="radio_tab_height">32dp</dimen>
+    <dimen name="card_corner_radius">6dp</dimen>
+    <dimen name="switch_widget_height">24dp</dimen>
```
done

omerhabib26 commented on OpenEdXMobile/res/values/styles.xml:
```diff
@@ -579,4 +580,10 @@
         <item name="android:textColor">@color/primaryBaseColor</item>
     </style>
+
+    <style name="sync.Calendar.Switch" parent="Theme.AppCompat.Light">
```
updated

omerhabib26 commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -238,27 +238,56 @@
+        If you happen to fall behind, you’ll be able to adjust the dates to keep yourself on track.</string>
+    <!-- Message Course Dates Banner Upgrade to Complete Graded -->
+    <string name="course_dates_banner_upgrade_to_graded">To complete graded assignments as part of this course, you can upgrade today.</string>
+    <!-- Message Course Dates Banner Upgrade to Reset -->
+    <string name="course_dates_banner_upgrade_to_reset">You are auditing this course,
+        which means that you are unable to participate in graded assignments. It looks like you missed some important deadlines based on our suggested schedule. To complete graded assignments as part of this course and shift the past due assignments into the future, you can upgrade today.</string>
+    <!-- Message Course Dates Banner Reset Date -->
+    <string name="course_dates_banner_reset_date">It looks like you missed some important deadlines based on our suggested schedule.
+        To keep yourself on track, you can update this schedule and shift the past due assignments into the future. Don’t worry—you won’t lose any of the progress you’ve made when you shift your due dates.</string>
```
On the new banner layout, we aren't following bold headings so that's why I added these

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -54,6 +63,16 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
     }
     private var courseData: EnrolledCoursesResponse = EnrolledCoursesResponse()
     private var isSelfPaced: Boolean = true
+    private var userType: String = ""
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -54,6 +63,16 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+
+    @Inject
+    private lateinit var loginPrefs: LoginPrefs
+    private val logger = Logger(javaClass.name)
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -87,10 +106,14 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
         viewModel = ViewModelProvider(this, ViewModelFactory()).get(CourseDateViewModel::class.java)
 
         courseData = arguments?.getSerializable(Router.EXTRA_COURSE_DATA) as EnrolledCoursesResponse
+        courseName = "${environment.config.platformName} - ${courseData.course.name}"
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CalendarUtils.kt:
```diff
@@ -0,0 +1,111 @@
+         * Method to check if the calendar with the course name exist in the mobile calendar app or not
+         */
+        @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+        fun getCalenderId(context: Context, accountName: String, courseName: String): Long {
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CalendarUtils.kt:
```diff
@@ -0,0 +1,111 @@
+         */
+        @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+        fun getCalenderId(context: Context, accountName: String, courseName: String): Long {
+            var calenderId = -1
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -87,10 +106,14 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+        courseName = "${environment.config.platformName} - ${courseData.course.name}"
         isSelfPaced = courseData.course.isSelfPaced
+        userType = courseData.mode
+        accountName = loginPrefs.currentUserProfile?.name ?: "local_user"
```
Done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/ConfigUtil.kt:
```diff
@@ -126,6 +131,36 @@ class ConfigUtil {
+            // Check firebase enabled in config
+            if (config.firebaseConfig.isEnabled) {
+                val firebaseRemoteConfig = FirebaseRemoteConfig.getInstance()
+                val configSettings = FirebaseRemoteConfigSettings.Builder()
+                        .setMinimumFetchIntervalInSeconds(10)
+                        .build()
+                firebaseRemoteConfig.setConfigSettingsAsync(configSettings)
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -196,4 +311,47 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+        calID = if (calID == (-1).toLong()) {
+            CalendarUtils.createCalendar(context = contextOrThrow, accountName = accountName, courseName = courseName)
+        } else {
+            deleteCalendar(calID)
```
Done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -167,6 +190,86 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
+                    ContentUris.appendId(builder, Calendar.getInstance().timeInMillis)
+                    val intent = Intent(Intent.ACTION_VIEW)
+                            .setData(builder.build())
+                    startActivity(intent)
```
Done

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CalendarUtils.kt:
```diff
@@ -0,0 +1,111 @@
+            context.contentResolver.delete(Uri.parse("content://com.android.calendar/calendars/$calendarId"), null, null)
+        }
+
+        private fun asSyncAdapter(uri: Uri, account: String): Uri? {
```
Added

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+    void trackCourseDatesCalendarToggleButton(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced, boolean toggle);
+
+    /**
+     * Track if the application has the read/write access of calendar or not
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+    void trackCourseDatesCalendarPermission(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced, boolean hasPermission);
+
+    /**
+     * Track whether the user confirm to allow the events to be added on calendar or not
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+    void trackCourseDatesViewCalendar(@NonNull String courseId, @NonNull String userType, @NonNull boolean isSelfPaced);
+
+    /**
+     * Track the user tapped on Done on Success dialog of the events added on calendar
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+     * @param courseId      Id of the course
+     * @param userType      Whether the Account type of the user is audit or verified
+     * @param isSelfPaced   Whether the course is self-paced or instructor paced
+     * @param hasPermission True if the application has permission of calendar otherwise False
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -551,9 +551,79 @@ void trackValuePropLearnMoreTapped(@NonNull String courseId, @Nullable String as
+     * @param courseId    Id of the course
+     * @param userType    Whether the Account type of the user is audit or verified
+     * @param isSelfPaced Whether the course is self-paced or instructor paced
+     * @param toggle      True if user allowed the events to be added on calendar otherwise False
```
removed

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CourseDateUtil.kt:
```diff
@@ -19,41 +21,55 @@ object CourseDateUtil {
-        val textView = view.findViewById(R.id.banner_info) as TextView
+        val containerLayout = view as LinearLayout
+        val title = view.findViewById(R.id.banner_title) as TextView
+        val description = view.findViewById(R.id.banner_info) as TextView
```
updated

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/ConfigUtil.kt:
```diff
@@ -90,19 +94,19 @@ class ConfigUtil {
-                                    listener.onCourseUpgradeResult(true)
-                                    return@addOnCompleteListener
-                                }
+                    val whiteListedReleasesJson = firebaseRemoteConfig
+                            .getString(AppConstants.FirebaseConstants.REV_934_WHITELISTED_RELEASES)
+                    if (!TextUtils.isEmpty(whiteListedReleasesJson)) {
+                        val whiteListedReleases = Gson().fromJson<ArrayList<String>>(whiteListedReleasesJson,
+                                object : TypeToken<ArrayList<String>>() {}.type)
+                        // Check current release is white listed in firebase remote config
+                        for (release in whiteListedReleases) {
+                            if (BuildConfig.VERSION_NAME.equals(release, ignoreCase = true)) {
+                                listener.onCourseUpgradeResult(true)
+                                return@addOnCompleteListener
                             }
                         }
+                    }
```
the current change is just the auto-formatting

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/CalendarUtils.kt:
```diff
@@ -0,0 +1,132 @@
+            fragment.startActivity(intent)
+        }
+    }
+}
```
need new line char at the end.

farhan-arshad-dev reviewed (approved):
Just a nit.

farhan-arshad-dev commented:
Verify the followings:

- [ ] An "add to calendar" message and CTA (switch) is shown on the Course Dates page
- [ ] Audit track users see the message, To complete graded assignments as part of this course, you can upgrade 
- [ ] Verified track users see a different message, We built a suggested schedule to help you stay on track. But don’t worry – it’s flexible so you can learn at your own pace. If you happen to fall behind, you’ll be able to adjust the dates to keep yourself on track.
- [ ] Message/CTA is shown only for self-paced, and hidden for instructor-paced courses
- [ ] Remote flags exist to show/hide the message and CTA
- [ ] Switching a CTA from off to on for the first time prompts the user for permission to modify their calendar
- [ ] If the user does not allow permission, switching a CTA from off to on as per Android platform standard experience.
- [ ] Switching the CTA from off to on from the Course Dates page presents a native alert or modal to confirm adding the course events:
	- Title: Add calendar “Technical Support Fundamentals”
	- Body: Would you like to add the edX calendar “<Calendar Name>”?You can edit or remove your course calendar any time in Calendar or Settings.
- [ ] Confirming adding course events adds all events found on the Course Dates page to the device's calendar. Events are visible in the calendar.
- [ ] Added course events
- [ ] Calendar events are added to a course specific calendar named edX - <Course Name>
- [ ] The calendar is created if it does not already exist
- [ ] Are 1 hour in duration
- [ ] Start 1 hour before the due time
- [ ] Have the name Assignment Due: <Course Name>
- [ ] Have a description with the same text as in the Dates tab, without the deep link (deep link is in a different story).  For example: “Homework 6 Peer Assessment (Submission)” or “MA1 section graded exam questions: MA1 Section 3 graded exam questions”
- [ ] Switching the CTA from on to off from the Course Dates page deletes the previously added events
- [ ] Analytics events implemented per specification.
